gh-103921: Document PEP 695 (#104642)
python/cpython: Doc/reference/compound_stmts.rst

   :token:`~python-grammar:assignment_expression`. Previously, the grammar was
   much more restrictive; see :pep:`614` for details.

A list of :ref:`type parameters <type-params>` may be given in square brackets
between the function's name and the opening parenthesis for its parameter list.
This indicates to static type checkers that the function is generic. At runtime,
the type parameters can be retrieved from the function's ``__type_params__``
attribute. See :ref:`generic-functions` for more.

> .. versionadded:: 3.12

Reviewer: Not sure how much it matters, but I would use `.. versionchanged:: 3.12` here, since we are describing function definition syntax in this section, and function definitions are not new in 3.12. The existing precedent of positional-only syntax below, which was "new" in Python 3.8, uses `versionchanged`.
```suggestion
.. versionchanged:: 3.12
```

>    Type parameters are new in version 3.12.

Reviewer 2: ```suggestion
   Type parameter lists are new in Python 3.12.
```

   :token:`~python-grammar:assignment_expression`. Previously, the grammar was
   much more restrictive; see :pep:`614` for details.

A list of :ref:`type parameters <type-params>` may be given in square brackets
immediately after the class's name.
This indicates to static type checkers that the class is generic. At runtime,
the type parameters can be retrieved from the class's ``__type_params__``
attribute. See :ref:`generic-classes` for more.

> .. versionadded:: 3.12

Reviewer: ```suggestion
.. versionchanged:: 3.12
```

>    Type parameters are new in version 3.12.

Reviewer 2: ```suggestion
   Type parameters are new in Python 3.12.
```

:ref:`Functions <def>` (including :ref:`coroutines <async def>`),
:ref:`classes <class>` and :ref:`type aliases <type>` may
contain a type parameter list::

   def max[T](args: list[T]) -> T:
       ...

   async def amax[T](args: list[T]) -> T:
       ...

   class Bag[T]:
       def __iter__(self) -> Iterator[T]:
           ...

       def add(self, arg: T) -> None:
           ...

   type ListOrSet[T] = list[T] | set[T]

Semantically, this indicates that the function, class, or type alias is
generic over a type variable. This information is primarily used by static
type checkers, and at runtime, generic objects behave much like their
non-generic counterparts.

Type parameters are declared in square brackets (``[]``) immediately
after the name of the function, class, or type alias. The type parameters
are accessible within the scope of the generic object, but not elsewhere.
Thus, after a declaration ``def func[T](): ...``, the name ``T`` is not in
the module scope. Below, the semantics of generic objects are described
with more precision. The scope of type parameters is modeled with a special
function (technically, an :ref:`annotation scope <annotation-scopes>`,

> represented with the pseudo-keyword ``def'``) that

Reviewer: Referring to the "pseudo-keyword def'" here, in the absence of any immediate translation using it, I think is more confusing than enlightening. I would leave out the mention of `def'` until the first actual example (below) that uses it.

the bound and constraints are not evaluated when the generic object is created,
but only when the value is explicitly accessed through the attributes ``__bound__``
and ``__constraints__``. To accomplish this, the bounds or constraints are
evaluated in a separate :ref:`annotation scope <annotation-scopes>`.

:data:`typing.TypeVarTuple`\ s and :data:`typing.ParamSpec`\ s cannot have bounds
or constraints.

The following example indicates the full set of allowed type parameter declarations::

   def overly_generic[
      SimpleTypeVar,
      TypeVarWithBound: int,
      TypeVarWithConstraints: (str, bytes),
      *SimpleTypeVarTuple,
      **SimpleParamSpec,
   ](
      a: SimpleTypeVar,
      b: TypeVarWithBound,
      c: Callable[SimpleParamSpec, TypeVarWithConstraints],
      *d: SimpleTypeVarTuple,
   ): ...

.. _generic-functions:

Generic functions
-----------------

Generic functions are declared as follows::

   def func[T](arg: T): ...

This syntax is equivalent to::

   def' TYPE_PARAMS_OF_func():
       T = typing.TypeVar("T")
       def func(arg: T): ...
       func.__type_params__ = (T,)
       return func
   func = TYPE_PARAMS_OF_func()

> Here ``def'`` indicates an :ref:`annotation scope <annotation-scopes>`. (One

Reviewer: I think it is worth a parenthetical aside to clarify that `def'` is not actually a keyword or part of Python, e.g. something like:
```suggestion
Here ``def'`` (not a real keyword) is used to indicate an :ref:`annotation scope <annotation-scopes>`. (One
```

> other liberty is taken in the translation: the syntax does not go through

Reviewer: Another way in which it is not equivalent is that nothing is ever actually bound to the name `TYPE_PARAMS_OF_func`. Perhaps this is obvious, but I think it is at least as relevant to mention as not going through attr access on typing module.

attribute access on the :mod:`typing` module, but creates an instance of
:data:`typing.TypeVar` directly.)

The annotations of generic functions are evaluated within the annotation scope
used for declaring the type parameters, but the function's defaults and
decorators are not.

The following example illustrates the scoping rules for these cases,
as well as for additional flavors of type parameters::

   @decorator
   def func[T: int, *Ts, **P](*args: *Ts, arg: Callable[P, T] = some_default):
       ...

> Except for the lazy evaluation of the ``TypeVar`` bound, this is equivalent to::

Reviewer: Link here to some fuller discussion of lazy evaluation.

Also, should `TypeVar` be a linked reference? I thought of this particularly for the benefit of someone who, reading this, needs a reminder what in the example is a "TypeVar bound."

   DEFAULT_OF_arg = some_default

>    def' TYPE_PARAMS_OF_func():

Reviewer: as above, I think it is worth clarifying, at least as an aside, that all the `*_OF_*` names are solely in order to illustrate the binding semantics, and no such name will ever exist

:ref:`classes <class>` and :ref:`type aliases <type>` may
contain a type parameter list::

   def max[T](args: list[T]) -> T:
       ...

   async def amax[T](args: list[T]) -> T:
       ...

   class Bag[T]:
       def __iter__(self) -> Iterator[T]:
           ...

       def add(self, arg: T) -> None:
           ...

   type ListOrSet[T] = list[T] | set[T]

Semantically, this indicates that the function, class, or type alias is
generic over a type variable. This information is primarily used by static
type checkers, and at runtime, generic objects behave much like their
non-generic counterparts.

Type parameters are declared in square brackets (``[]``) immediately
after the name of the function, class, or type alias. The type parameters
are accessible within the scope of the generic object, but not elsewhere.
Thus, after a declaration ``def func[T](): pass``, the name ``T`` is not available in
the module scope. Below, the semantics of generic objects are described
with more precision. The scope of type parameters is modeled with a special
function (technically, an :ref:`annotation scope <annotation-scopes>`) that
wraps the creation of the generic object.

> Generic functions, classes, and type aliases have a :attr:`__type_params__`

Reviewer 2: The link on this line isn't rendering properly: https://cpython-previews--104642.org.readthedocs.build/en/104642/reference/compound_stmts.html#type-parameter-lists

Author: Unlinked it

Reviewer 2: Thanks!

and ``__constraints__``. To accomplish this, the bounds or constraints are
evaluated in a separate :ref:`annotation scope <annotation-scopes>`.

:data:`typing.TypeVarTuple`\ s and :data:`typing.ParamSpec`\ s cannot have bounds
or constraints.

The following example indicates the full set of allowed type parameter declarations::

   def overly_generic[
      SimpleTypeVar,
      TypeVarWithBound: int,
      TypeVarWithConstraints: (str, bytes),
      *SimpleTypeVarTuple,
      **SimpleParamSpec,
   ](
      a: SimpleTypeVar,
      b: TypeVarWithBound,
      c: Callable[SimpleParamSpec, TypeVarWithConstraints],
      *d: SimpleTypeVarTuple,
   ): ...

.. _generic-functions:

Generic functions
-----------------

Generic functions are declared as follows::

   def func[T](arg: T): ...

This syntax is equivalent to::

   annotation-def TYPE_PARAMS_OF_func():
       T = typing.TypeVar("T")
       def func(arg: T): ...
       func.__type_params__ = (T,)
       return func
   func = TYPE_PARAMS_OF_func()

Here ``annotation-def`` indicates an :ref:`annotation scope <annotation-scopes>`,
which is not actually bound to any name at runtime. (One

> other liberties are taken in the translation: the syntax does not go through

Reviewer 2: grammar :)

```suggestion
other liberty is taken in the translation: the syntax does not go through
```

.. index::
   single: type parameters

.. productionlist:: python-grammar
   type_params: "[" `type_param` ("," `type_param`)* "]"
   type_param: `typevar` | `typevartuple` | `paramspec`
   typevar: `identifier` (":" `expression`)?
   typevartuple: "*" `identifier`
   paramspec: "**" `identifier`

:ref:`Functions <def>` (including :ref:`coroutines <async def>`),
:ref:`classes <class>` and :ref:`type aliases <type>` may
contain a type parameter list::

   def max[T](args: list[T]) -> T:
       ...

   async def amax[T](args: list[T]) -> T:
       ...

   class Bag[T]:
       def __iter__(self) -> Iterator[T]:
           ...

       def add(self, arg: T) -> None:
           ...

   type ListOrSet[T] = list[T] | set[T]

Semantically, this indicates that the function, class, or type alias is
generic over a type variable. This information is primarily used by static
type checkers, and at runtime, generic objects behave much like their
non-generic counterparts.

Type parameters are declared in square brackets (``[]``) immediately
after the name of the function, class, or type alias. The type parameters
are accessible within the scope of the generic object, but not elsewhere.

> Thus, after a declaration ``def func[T](): ...``, the name ``T`` is not in

Reviewer 2: People who don't write stubs on a daily basis are probably more familiar with `pass` :p

```suggestion
Thus, after a declaration ``def func[T](): pass``, the name ``T`` is not available in
```

the module scope. Below, the semantics of generic objects are described
with more precision. The scope of type parameters is modeled with a special
function (technically, an :ref:`annotation scope <annotation-scopes>`) that
wraps the creation of the generic object.

Generic functions, classes, and type aliases have a :attr:`__type_params__`
attribute listing their type parameters.

Type parameters come in three kinds:

* :data:`typing.TypeVar`, introduced by a plain name (e.g., ``T``). Semantically, this

>   stands for a single type.

Reviewer 2: ```suggestion
  represents a single type to a static type checker.
```

Author: Changed but left out "static", it works for dynamic type checkers too.

Reviewer 2: #JusticeForPyanalyze

* :data:`typing.TypeVarTuple`, introduced by a name prefixed with a single
  asterisk (e.g., ``*Ts``). Semantically, this stands for a tuple of any
  number of types.
* :data:`typing.ParamSpec`, introduced by a name prefixed with two asterisks
  (e.g., ``**P``). Semantically, this stands for the parameters of a callable.

:data:`typing.TypeVar` declarations can define *bounds* and *constraints* with
a colon (``:``) followed by an expression. A single expression after the colon
indicates a bound (e.g. ``T: int``). The expression should be a type (though this

> is not enforced at runtime, only by static type checkers) and semantically, this means

Reviewer 2: Not sure this sentence is useful:

> The expression should be a type (though this is not enforced at runtime, only by static type checkers)

To me, it immediately begs the question, "But what is a type?", and that's not such an easy thing to answer from a static-typing perspective. `TypeVar`s can be bound to classes, but `int | str`, `Union[int, str]` and `list[int]` are also all valid bounds. Meanwhile, `Final[str]` is not a valid bound, and neither is a `ParamSpec`.

I wonder how important it is here to go in depth into what users _should_ be doing with this syntax if they want it to be understood by static type checkers. This section of the docs is a technical reference manual for the way the syntax works. The static-typing implications are almost irrelevant here.

Author: I removed the first part, but I do want to keep some indication of what this syntax is meant for, so readers of the docs aren't left wondering what this is all about.

take on one of the types in the list of constraints.

For :data:`!typing.TypeVar`\ s declared using the type parameter list syntax,
the bound and constraints are not evaluated when the generic object is created,
but only when the value is explicitly accessed through the attributes ``__bound__``
and ``__constraints__``. To accomplish this, the bounds or constraints are
evaluated in a separate :ref:`annotation scope <annotation-scopes>`.

:data:`typing.TypeVarTuple`\ s and :data:`typing.ParamSpec`\ s cannot have bounds
or constraints.

The following example indicates the full set of allowed type parameter declarations::

   def overly_generic[
      SimpleTypeVar,
      TypeVarWithBound: int,
      TypeVarWithConstraints: (str, bytes),
      *SimpleTypeVarTuple,
      **SimpleParamSpec,
   ](
      a: SimpleTypeVar,
      b: TypeVarWithBound,
      c: Callable[SimpleParamSpec, TypeVarWithConstraints],
      *d: SimpleTypeVarTuple,
   ): ...

.. _generic-functions:

Generic functions
-----------------

Generic functions are declared as follows::

   def func[T](arg: T): ...

This syntax is equivalent to::

>    def' TYPE_PARAMS_OF_func():

Reviewer 2: I think `def'` is maybe a little too subtle -- it looks like it could be a typo.

Maybe `scopedef`?

Author: Yes, it looks like grit on Tim Peters' monitor :)

I used `def695` in the PEP itself, but I don't want to tie us too closely to the PEP number.

"scopedef" is weird to me because normal functions make a scope too. Maybe "annotation-def" to tie us in with the name "annotation scopes".

Reviewer 2: annotation-def is probably the clearest, yeah. It's the most verbose — but you can't have everything :)

Author: I don't think verbose is bad here, it makes it clear there's something unusual going on and it's not just a typo.

Author: Switched to `annotation-def`.
